nodes.go

```diff
@@ -412,6 +412,31 @@ func (l *literal) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.
 	return nil, nil
 }
 
+type negation struct {
+	node node
+}
+
+func (n *negation) String() string { return "!" + stringer(n.node) }
+
+func (n *negation) Parse(ctx *parseContext, parent reflect.Value) (out []reflect.Value, err error) {
+	// We create a clone to go back to the previous lexer
+	clone := newParseContext(ctx.Clone(), ctx.lookahead, ctx.caseInsensitive)
```

Comment:
Use `ctx.Branch()` and `ctx.Accept()`/`ctx.Stop()` for this conditional branching. See [optional](https://github.com/alecthomas/participle/blob/master/nodes.go#L330-L345) for an example.